draft-ietf-httpbis-rfc6265bis.md

```diff
@@ -1236,6 +1276,15 @@ user agent MUST process the cookie-av as follows.
 6.  Append an attribute to the cookie-attribute-list with an attribute-name
     of Expires and an attribute-value of expiry-time.
 
+Based on the parsing algorithm above, the syntax requirements for the Expires
+attribute are represented by the grammar below. Expires attributes not
+conforming to this grammar are ignored.
+
+~~~ abnf
+expires-av = "Expires" BWS "=" BWS cookie-date BWS
+               ; cookie-date is defined in the "Dates" section.
+~~~
```

Comment:
At the "Expires" is detected, the parsing into name and value has already happened.

Thus, I would just define the syntax for the value, as in:

~~~
expires-v = cookie-date
~~~

Note this is how multi-level grammars are defined in the core HTTP specs as well.

Reply:
Ah, I see - thank you for the example @reschke.

It seems like we'd lose a bit of completeness if we only include the attribute value in the grammar.  For instance, the grammar wouldn't explicitly tie the "Expires" keyword to the Expires attribute, and wouldn't convey whether "ExPiReS" is treated the same way as "Expires".  Having the grammar format used here align with the convention in the core HTTP specs seems desirable as well, though.

Does anyone else have thoughts on this?

Reply:
> . For instance, the grammar wouldn't explicitly tie the "Expires" keyword to the Expires attribute

It by definition can't do that anyway, as long as the grammar allows for extensibilty - an ABNF-driven parser would simply switch to the code branch for new attributes, and then the value syntax would remain unconstrained.

You really can't do this in ABNF; it needs to be done in prose.

Reply:
If our goal is to have an ABNF-driven parser function correctly, it seems like we'd need to go back to the using the previous definition of `cookie-av`. To address the original concern of "any valid cookie-av will match extension-av", maybe we could do something like the following:
```
cookie-av          = expires-av / max-age-av / domain-av /
                     path-av / secure-av / httponly-av /
                     samesite-av / extension-av

extension-av       = 1*cookie-name-octet BWS extension-eq-value BWS
extension-eq-value = "" / ("=" BWS optional-value)
    ; attributes solely matching extension-av should be treated as unrecognized by
    ; cookie parsers strictly conforming to this RFC
```

What do you think?

Reply:
I don't think you can solve this by adding prose (comments) to the ABNF.

The ABNF should be something that can translate to how a parser processes the input. Parsers in general parse first generic structures, then look at specific micro syntaxes of values.

That's what we're doing in the base spec, and that's what I'll keep suggesting to do.